### PR TITLE
Object Lambda docs

### PR DESCRIPTION
**Pull request by feorlen**
Object Lambda functions to transform objects via a webhook. Expand and refine the developer docs from

https://github.com/minio/minio/blob/c4d71bc1a52c8bf5d433b2c39b94b2408884024b/docs/lambda/README.md

into a new page in the user documentation.

Staged
http://192.241.195.202:9000/staging/DOCS-754/linux/html/developers/transforms-with-object-lambda.html

Notes for reviewers:

In addition to the items noted in inline comments, some general things to confirm:
* Accurate usage/terminology? Is this how we talk about the relevant topics?
* Any technical points that need more/more correct detail?
* What would someone need to know to go from trying this dev example to having a real function in their production environment?

[x] Add new docs page in Developers
[x] Add envars to envar list

Fixes https://github.com/minio/docs/issues/754

### CODE REVIEW

File: source/developers/transform-data-with-object-lambda.rst [hunk 1]
```diff
@@ -0,0 +1,242 @@
+.. _developers-object-lambda:
+
+=================================
+Transform Data with Object Lambda
```
Comment (feorlen) [inline source/developers/transform-data-with-object-lambda.rst:4]: Actual title tbd, recommendations welcome

File: source/developers/transform-data-with-object-lambda.rst [hunk 2]
```diff
@@ -0,0 +1,242 @@
...
+
+- An existing :ref:`MinIO <minio-installation>` deployment
+- Working Python (3.8+) and Golang (??) development environments
+- :doc:`The MinIO Go SDK </developers/go/minio-go>` installed
+  
+-------
+Example
+-------
+
+This example creates a Lambda handler in Python, invokes it from a Go function, and retrieves the transformed object with ``curl``.
```
Comment (feorlen) [inline source/developers/transform-data-with-object-lambda.rst:26]: Could the handler be something besides Python? What are other ways to access the transformed object?

Also, how should this work in prod?

File: source/developers/transform-data-with-object-lambda.rst [hunk 3]
```diff
@@ -0,0 +1,242 @@
...
+
+
+Create a Lambda Handler
+~~~~~~~~~~~~~~~~~~~~~~~~~
+
+The sample handler, written in Python, retrieves the target object using a `presigned URL <https://min.io/docs/minio/linux/developers/go/API.html#presigned-operations>`__. It then transforms the object's contents and returns the new text to the caller.
```
Comment (feorlen) [inline source/developers/transform-data-with-object-lambda.rst:32]: "returns the new text to the caller" is vague. What is a better way to describe the path back to who/what requested the transformed object?

File: source/developers/transform-data-with-object-lambda.rst [hunk 4]
```diff
@@ -0,0 +1,242 @@
...
+.. code-block:: shell
+   :class: copyable
+
+   pip install flask requests
+
+The handler uses information from the MinIO event context to get the original object, along with the tokens needed to validate the response when the transformed object is returned.
+In the example below, ``object_context`` contains several values from the ``getObjectContext`` property of the JSON request payload:
+
+* ``inputS3Url`` is a presigned URL for the original object. A `presigned URL <https://min.io/docs/minio/linux/developers/go/API.html#presigned-operations>`__ allows the function to access an object without the MinIO credentials usually required. 
+* ``outputRoute`` and ``outputToken`` are tokens added to the response headers to validate the response and return it to the correct user.
```
Comment (feorlen) [inline source/developers/transform-data-with-object-lambda.rst:46]: More vagueness, particularly about `outputRoute`. I presume that is how we know who to send the response back to, what additional can we say about that?

File: source/developers/transform-data-with-object-lambda.rst [hunk 5]
```diff
@@ -0,0 +1,242 @@
...
+
+         # Get the original S3 object using the presigned URL
+         r = requests.get(s3_url)
+         original_object = r.content.decode('utf-8')
+
+         # Transform all text in the original object to uppercase
+         # You can replace it with your custom code based on your use case
+         transformed_object = original_object.upper()
+
+         # Write object back to S3 Object Lambda
```
Comment (feorlen) [inline source/developers/transform-data-with-object-lambda.rst:83]: It seems the results are ephemeral. Is the presigned URL one time use?

File: source/developers/transform-data-with-object-lambda.rst [hunk 6]
```diff
@@ -0,0 +1,242 @@
...
+         resp.headers['x-amz-request-token'] = request_token
+         return resp
+
+      else:
+         abort(400)
+
+   if __name__ == '__main__':
+      app.run()
+
+You may perform any actions needed for your use case, including saving the transformed object.
```
Comment (feorlen) [inline source/developers/transform-data-with-object-lambda.rst:97]: What does someone need to know about allowed actions inside the handler?

File: source/developers/transform-data-with-object-lambda.rst [hunk 7]
```diff
@@ -0,0 +1,242 @@
...
+
+Start the Handler
++++++++++++++++++
+
+To test the handler in your local development environment, start it with the following command:
+
+.. code-block:: shell
+   :class: copyable
+
+   python lambda_handler.py
```
Comment (feorlen) [inline source/developers/transform-data-with-object-lambda.rst:109]: ehhh for me it's `python3` but I think that's sufficiently well understood (by Python people) that I'm going to gloss over it. (Same for venv.)

File: source/developers/transform-data-with-object-lambda.rst [hunk 8]
```diff
@@ -0,0 +1,242 @@
...
+    * Running on http://127.0.0.1:5000
+   Press CTRL+C to quit
+
+Once the handler is running, start MinIO with the :envvar:`MINIO_LAMBDA_WEBHOOK_ENABLE` and :envvar:`MINIO_LAMBDA_WEBHOOK_ENDPOINT` environment variables to register the function with MinIO.
+To identify the specific Object Lambda handler, append the name of the function to the name of the environment variable:
+
+.. code-block:: shell
+   :class: copyable
+
+   MINIO_LAMBDA_WEBHOOK_ENABLE_myfunction=on MINIO_LAMBDA_WEBHOOK_ENDPOINT_myfunction=http://localhost:5000 minio server /data &
```
Comment (feorlen) [inline source/developers/transform-data-with-object-lambda.rst:127]: This starts the process in the background, what is the correct way to stop it?

File: source/developers/transform-data-with-object-lambda.rst [hunk 9]
```diff
@@ -0,0 +1,242 @@
...
+         if err != nil {
+            log.Fatalln(err)
+         }
+
+         // Set lambda function target via `lambdaArn`
+         reqParams := make(url.Values)
+         reqParams.Set("lambdaArn", "arn:minio:s3-object-lambda::myfunction:webhook")
+
+         // Generate presigned GET url with lambda function
+         presignedURL, err := s3Client.PresignedGetObject(context.Background(), "myfunctionbucket", "testobject", time.Duration(1000)*time.Second, reqParams)
```
Comment (feorlen) [inline source/developers/transform-data-with-object-lambda.rst:198]: Oops need some text noting where to put your own function, bucket, and object names.

File: source/developers/transform-data-with-object-lambda.rst [hunk 10]
```diff
@@ -0,0 +1,242 @@
...
+      curl -v $(go run presigned.go)
+
+   The test function generates the presigned URL, passing it to ``curl`` to access the object. 
+   The output resembles the following:
+
+   .. code-block:: shell
+
+      *   Trying 127.0.0.1:9000...
+      * Connected to localhost (127.0.0.1) port 9000 (#0)
+      > GET /myfunctionbucket/testobject?X-Amz-Algorithm=AWS4-HMAC-SHA256&X-Amz-Credential=minioadmin%2F20230406%2Fus-east-1%2Fs3%2Faws4_request&X-Amz-Date=20230406T184749Z&X-Amz-Expires=1000&X-Amz-SignedHeaders=host&lambdaArn=arn%3Aminio%3As3-object-lambda%3A%3Amyfunction%3Awebhook&X-Amz-Signature=68fe7e03929a7c0da38255121b2ae09c302840c06654d1b79d7907d942f69915 HTTP/1.1
```
Comment (feorlen) [inline source/developers/transform-data-with-object-lambda.rst:219]: Is anything in here sensitive? Or it doesn't matter because it expires quickly?

File: source/index.rst [hunk 11]
```diff
@@ -180,6 +180,7 @@ Any file uploaded to ``play`` should be considered public and non-protected.
 
       Software Development Kits (SDK) <https://min.io/docs/minio/linux/developers/minio-drivers.html?ref=docs>
       Security Token Service (STS) <https://min.io/docs/minio/linux/developers/security-token-service.html?ref=docs>
+      Object Lambda <https://min.io/docs/minio/linux/developers/transform-data-with-object-lambda.html?ref=docs>
```
Comment (feorlen) [inline source/index.rst:183]: The title here appears to be replaced with the page title from the file, I guess this one should match?

File: source/reference/minio-server/minio-server.rst [hunk 12]
```diff
@@ -528,11 +528,11 @@ server logs webhook endpoints:
 
    export MINIO_LOGGER_WEBHOOK_ENABLE_PRIMARY="on"
    export MINIO_LOGGER_WEBHOOK_AUTH_TOKEN_PRIMARY="TOKEN"
-   export MINIO_LOGGER_WEBHOOK_ENDPOINT_PRIMARY="http://webhook-1.example.net
+   export MINIO_LOGGER_WEBHOOK_ENDPOINT_PRIMARY="http://webhook-1.example.net"
```
Comment (feorlen) [inline source/reference/minio-server/minio-server.rst:531]: One of several drive-by fixes for trailing quotes elsewhere in the page.

File: source/reference/minio-server/minio-server.rst [hunk 13]
```diff
@@ -2198,6 +2198,39 @@ endpoints as ``PRIMARY`` and ``SECONDARY`` respectively:
    :mc-conf:`notify_webhook comment <notify_webhook.comment>`
    configuration setting.
 
+.. _minio-server-envvar-object-lambda-webhook:
```
Comment (feorlen) [inline source/reference/minio-server/minio-server.rst:2201]: This name could be better

File: source/reference/minio-server/minio-server.rst [hunk 14]
```diff
@@ -2198,6 +2198,39 @@ endpoints as ``PRIMARY`` and ``SECONDARY`` respectively:
...
+
+Object Lambda
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+The following section documents environment variables for configuring MinIO to
+publish data to an HTTP webhook endpoint for an Object Lambda function. See
```
Comment (feorlen) [inline source/reference/minio-server/minio-server.rst:2207]: "publish data" is probably not the most accurate description. Ideas? 

(Also these lines need to be joined into complete sentences.)

File: source/reference/minio-server/minio-server.rst [hunk 15]
```diff
@@ -2198,6 +2198,39 @@ endpoints as ``PRIMARY`` and ``SECONDARY`` respectively:
...
+.. envvar:: MINIO_LAMBDA_WEBHOOK_ENABLE
+
+   Specify ``"on"`` to enable publishing :mc:`minio server` logs to the Object Lambda
+   webhook endpoint for the specified function.
+
+   Requires specifying :envvar:`MINIO_LAMBDA_WEBHOOK_ENDPOINT`.
+
+.. envvar:: MINIO_LAMBDA_WEBHOOK_ENDPOINT
+
+   The HTTP endpoint of the webhook for the specified function.
```
Comment (feorlen) [inline source/reference/minio-server/minio-server.rst:2232]: If `MINIO_LAMBDA_WEBHOOK_ENABLE` requires `MINIO_LAMBDA_WEBHOOK_ENDPOINT`, is the reverse true? (Or maybe it's fine to define but not enable.)

File: source/developers/transform-data-with-object-lambda.rst (hunk 4 above)
Comment (djwfyi) [reply source/developers/transform-data-with-object-lambda.rst:46]: Maybe
```
* MinIO Lambda adds `outputRoute`` and ``outputToken`` to the response headers to validate the response and return it to the correct user.
```

File: source/developers/transform-data-with-object-lambda.rst (hunk 7 above)
Comment (djwfyi) [reply source/developers/transform-data-with-object-lambda.rst:109]: Maybe
> Use the following command to start the handler for testing in your local development environment:

Comment (djwfyi) [reply source/developers/transform-data-with-object-lambda.rst:109]: Also, yeah, just `python` is fine.

File: source/reference/minio-server/minio-server.rst (hunk 14 above)
Comment (djwfyi) [reply source/reference/minio-server/minio-server.rst:2207]: "furnish data"? 
"provide data"?

File: source/reference/minio-server/minio-server.rst [hunk 16]
```diff
@@ -2216,9 +2216,20 @@ For example, the following command set two distinct Object Lambda webhook endpoi
...
-   export MINIO_LAMBDA_WEBHOOK_ENDPOINT_myfunction="http://webhook-1.example.net
+   export MINIO_LAMBDA_WEBHOOK_ENDPOINT_myfunction="http://webhook-1.example.net"
    export MINIO_LAMBDA_WEBHOOK_ENABLE_yourfunction="on"
-   export MINIO_LAMBDA_WEBHOOK_ENDPOINT_yourfunction="http://webhook-2.example.net
+   export MINIO_LAMBDA_WEBHOOK_ENDPOINT_yourfunction="http://webhook-2.example.net"
+
+.. envvar:: MINIO_LAMBDA_WEBHOOK_ENABLE
+
+   Specify ``"on"`` to enable publishing :mc:`minio server` logs to the Object Lambda
+   webhook endpoint for the specified function.
```
Comment (djwfyi) [inline source/reference/minio-server/minio-server.rst:2226]: Is this just server logs that get output?

File: source/reference/minio-server/minio-server.rst (hunk 15 above)
Comment (djwfyi) [reply source/reference/minio-server/minio-server.rst:2232]: In other places, you can define something and then enable/disable it separately. But you can only enable it if it is defined.

**Review by djwfyi — CHANGES_REQUESTED** [review]
Good first pass.
I'd like to see more separation of "what this does" from the example. So have a chunk at the beginning where you talk more about what Object Lambda can do and the steps required to do something with it (create a handler, enable the environment variables, trigger an event).

Then let the example just be an example without trying to use the example to teach or instruct as you go, if that makes sense.

So have a structure like:

```
=====
TITLE
=====

.. frontmatter stuff

two-sentence intro paragraph

Description
------------

more detailed description of what this is with some expanded descriptions of what could be done with it (no details of how).

Getting Started Overview
-------------------------

1. Create handler
2. Start/Restart MinIO with the environment variables
3. Trigger a webhook event
4. and whatever other steps I missed

Example
--------

Example Title
~~~~~~~~~~~~~

Trimmed down example just doing the things with minimal explanation except for what is unique/specific this this specific example.

Example Title 2
~~~~~~~~~~~~~~~

Maybe not with this PR, but it'll be nice to have a second, different example down the road of another way to  use this feature.

```

File: source/developers/transform-data-with-object-lambda.rst (hunk 1 above)
Comment (feorlen) [reply source/developers/transform-data-with-object-lambda.rst:4]: I like that, thanks!

File: source/developers/transform-data-with-object-lambda.rst (hunk 7 above)
Comment (feorlen) [reply source/developers/transform-data-with-object-lambda.rst:109]: I simplified the sentence and bypassed the awkward wording entirely. Good idea.

File: source/developers/transform-data-with-object-lambda.rst (hunk 8 above)
Comment (feorlen) [reply source/developers/transform-data-with-object-lambda.rst]: I avoided the whole thing by removing `&` from the example. There's no reason it has to be a background process for testing.

File: source/developers/transform-data-with-object-lambda.rst (hunk 6 above)
Comment (harshavardhana) [reply source/developers/transform-data-with-object-lambda.rst]: That is not generally our problem, that is what the above sentence wishes to convey.

The handler is their implementation we do not dictate on what they should do in that.

They may do string manipulation, they made to virus scanning, they may do PII verification of the document being read.. and block it if suspicious.

There are numerous ways to do it, all we need to talk about is how to use such a lambda function from MinIO and what MinIO expects as a response from this function handler.

As we need a response to respond back to the client that asked for transformation of their data.

Comment (feorlen) [reply source/developers/transform-data-with-object-lambda.rst]: I've revised some of the text to focus more on mechanics of handler functions.

File: source/reference/minio-server/minio-server.rst (hunk 14 above)
Comment (feorlen) [reply source/reference/minio-server/minio-server.rst]: Fiddled with the text but kept `publish` after all. It seems to be a common way to describe sending data to a webhook endpoint.

File: source/developers/transform-data-with-object-lambda.rst [hunk 17]
```diff
@@ -0,0 +1,242 @@
...
+
+The following documentation covers transforming the contents of an object with MinIO's Object Lambda.
+Uses for Object Lambda include redacting personally identifable information (PII) and enriching data with information from other sources.
+
+This documentation assumes the following:
+
+- An existing :ref:`MinIO <minio-installation>` deployment
+- Working Python (3.8+) and Golang (??) development environments
+- :doc:`The MinIO Go SDK </developers/go/minio-go>` installed
+
```
[the remark this answers is not in the document]
Comment (feorlen) [reply source/developers/transform-data-with-object-lambda.rst]: Added more discussion in several locations.

File: source/reference/minio-server/minio-server.rst (hunk 16 above)
Comment (feorlen) [reply source/reference/minio-server/minio-server.rst]: Dropped the logs bit, as there don't appear to be any involved. Just the request/response for the handler.

File: source/developers/transform-data-with-object-lambda.rst [hunk 18]
```diff
@@ -0,0 +1,242 @@
...
+- :doc:`The MinIO Go SDK </developers/go/minio-go>` installed
+  
+-------
+Example
+-------
+
+This example creates a Lambda handler in Python, invokes it from a Go function, and retrieves the transformed object with ``curl``.
+
+
+Create a Lambda Handler
```
[the remark this answers is not in the document]
Comment (feorlen) [reply source/developers/transform-data-with-object-lambda.rst]: Reworked example section throughout

File: source/developers/transform-data-with-object-lambda.rst [hunk 19]
```diff
@@ -0,0 +1,242 @@
...
+~~~~~~~~~~~~~~~~~~~~~~~~~
+
+The sample handler, written in Python, retrieves the target object using a `presigned URL <https://min.io/docs/minio/linux/developers/go/API.html#presigned-operations>`__. It then transforms the object's contents and returns the new text to the caller.
+It uses `the Flask web framework <https://flask.palletsprojects.com/en/2.2.x/>`__ and Python 3.8+.
```
[the remark this answers is not in the document]
Comment (feorlen) [reply source/developers/transform-data-with-object-lambda.rst]: Good idea, went back to "handler" here.

File: source/developers/transform-data-with-object-lambda.rst [hunk 20]
```diff
@@ -0,0 +1,242 @@
...
+
+The handler uses information from the MinIO event context to get the original object, along with the tokens needed to validate the response when the transformed object is returned.
+In the example below, ``object_context`` contains several values from the ``getObjectContext`` property of the JSON request payload:
+
+* ``inputS3Url`` is a presigned URL for the original object. A `presigned URL <https://min.io/docs/minio/linux/developers/go/API.html#presigned-operations>`__ allows the function to access an object without the MinIO credentials usually required. 
+* ``outputRoute`` and ``outputToken`` are tokens added to the response headers to validate the response and return it to the correct user.
+
+The handler calls string ``upper()`` to convert the text to all uppercase letters.
+It then sends the results back to MinIO, which returns it to the caller.
+The original object is not changed.
```
[the remark this answers is not in the document]
Comment (feorlen) [reply source/developers/transform-data-with-object-lambda.rst]: Added additional text about original vs transformed, since the whole point is one original and potentially infinite transformed alternatives.

File: source/developers/transforms-with-object-lambda.rst [hunk 21]
```diff
@@ -0,0 +1,353 @@
...
+
+.. code-block:: shell
+
+    * Serving Flask app 'lambda_handler'
+    * Debug mode: off
+   WARNING: This is a development server. Do not use it in a production deployment. Use a production WSGI server instead.
+    * Running on http://127.0.0.1:5000
+   Press CTRL+C to quit
+
+Once the handler is running, start MinIO with the :envvar:`MINIO_LAMBDA_WEBHOOK_ENABLE` and :envvar:`MINIO_LAMBDA_WEBHOOK_ENDPOINT` environment variables to register the function with MinIO.
```
Comment (djwfyi) [inline source/developers/transforms-with-object-lambda.rst:223]: Maybe swap the order.
Set the environment variables, then start MinIO.

File: source/developers/transforms-with-object-lambda.rst [hunk 22]
```diff
@@ -0,0 +1,353 @@
...
+            log.Fatalln(err)
+         }
+	 
+         // Print the URL to stdout
+         fmt.Println(presignedURL)
+      }      
+
+   In the code above, replace the following values:
+
+   - Replace ``my_admin_user`` and ``my_admin_password`` with root credentials for a MinIO deployment
```
Comment (djwfyi) [inline source/developers/transforms-with-object-lambda.rst:313]: Can you use a non-root credential to do transforms?

**Review by djwfyi — CHANGES_REQUESTED** [review]
Mostly minor suggestions to switch from passive to active and follow ventilated prose style.

This is a great work, though.
I'll take one more quick look through after this, but it's very close to ready!

File: source/developers/transforms-with-object-lambda.rst [hunk 23]
```diff
@@ -0,0 +1,353 @@
...
+MinIO's Object Lambda enables developers to programmatically transform objects on demand.
+You can transform objects in any way needed for your use case, such as redacting personally identifiable information (PII), enriching data with information from other sources, or converting between formats.
+
+Overview
+--------
+
+An Object Lambda handler is a small code module that transforms the contents of an object and returns the results.
+Like :s3-docs:`Amazon S3 Object Lambda functions <transforming-objects.html>`, a MinIO Object Lambda handler function is triggered by a GET request from an application.
+The handler retrieves the requested object from MinIO, transforms it, and returns the modified data back to MinIO to send to the original application.
+The original object is unchanged.
```
[the remark this answers is not in the document]
Comment (feorlen) [reply source/developers/transforms-with-object-lambda.rst:22]: Yeah that works!

File: source/developers/transforms-with-object-lambda.rst (hunk 22 above)
Comment (feorlen) [reply source/developers/transforms-with-object-lambda.rst:313]: Hmm. No policy fails, as expected. And `readwrite` succeeds. But `readonly` is also ok?
```
$ mc admin policy attach myminio readonly --user=feorlen
Policy `readonly` successfully attached to user `feorlen`
$ curl -v $(go run presigned.go)
[...]
hELLO, wORLD!
```

@harshavardhana what permissions should/should not work for triggering Object Lambda?

Comment (harshavardhana) [reply source/developers/transforms-with-object-lambda.rst:313]: yes you are only doing "readonly" there is no WRITE anywhere in the code in lambda layer.

File: source/developers/transform-data-with-object-lambda.rst [hunk 24]
```diff
@@ -0,0 +1,242 @@
...
+It uses `the Flask web framework <https://flask.palletsprojects.com/en/2.2.x/>`__ and Python 3.8+. 
+
+The following command installs Flask and other needed dependancies:
+
+.. code-block:: shell
+   :class: copyable
+
+   pip install flask requests
+
+The handler uses information from the MinIO event context to get the original object, along with the tokens needed to validate the response when the transformed object is returned.
```
[the remark this answers is not in the document]
Comment (feorlen) [reply source/developers/transform-data-with-object-lambda.rst]: Addressed as part of other edits.

File: source/developers/transforms-with-object-lambda.rst [hunk 25]
```diff
@@ -0,0 +1,353 @@
...
+   .. list-table::
+      :widths: 25 75
+      :header-rows: 1
+
+      * - Value
+        - Description
+
+      * - ``inputS3Url``
+        - A `presigned URL <https://min.io/docs/minio/linux/developers/go/API.html#presigned-operations>`__ for the original object. This URL allows the handler to access an object without the MinIO credentials usually required.
+          This URL is generated by the calling application and is valid for one hour after it is created.
```
[the remark this answers is not in the document]
Comment (feorlen) [reply source/developers/transforms-with-object-lambda.rst:55]: Ended up rewriting the entire list item a different way.

File: source/developers/transforms-with-object-lambda.rst (hunk 21 above)
Comment (feorlen) [reply source/developers/transforms-with-object-lambda.rst:223]: For this purpose, it's the correct order because you are running in a local shell. Added another heading and additional text to clarify you are starting MinIO for testing, with these environment variables. Maybe that is easier to follow. Something like this:

> Start MinIO
> 
> (envvar detail)
> 
> The following command starts MinIO in your local development environment:
> 
> (code block)

Comment (feorlen) [reply source/developers/transforms-with-object-lambda.rst:234]: @djwfyi Here's the relevant section in the restaged version
http://192.241.195.202:9000/staging/DOCS-754/linux/html/developers/transforms-with-object-lambda.html#start-the-handler

**Review by djwfyi — APPROVED** [review]
Nice work!
LGTM
